File: Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/TextToolCommand.java
/*
 * Paintroid: An image manipulation application for Android.
 * Copyright (C) 2010-2015 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.command.implementation;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;
import android.graphics.Rect;

import org.catrobat.paintroid.command.Command;
import org.catrobat.paintroid.contract.LayerContracts;

public class TextToolCommand implements Command {
	private final String[] multilineText;
	private final Paint textPaint;
	private final float boxOffset;
	private final float boxWidth;
	private final float boxHeight;
	private final PointF toolPosition;
	private final float rotationAngle;

	public TextToolCommand(String[] multilineText, Paint textPaint, float boxOffset,
			float boxWidth, float boxHeight, PointF toolPosition, float rotationAngle) {
		this.multilineText = multilineText.clone();
		this.textPaint = textPaint;
		this.boxOffset = boxOffset;
		this.boxWidth = boxWidth;
		this.boxHeight = boxHeight;
		this.toolPosition = toolPosition;
		this.rotationAngle = rotationAngle;
	}

	@Override
	public void run(Canvas canvas, LayerContracts.Model layerModel) {
		canvas.save();

		canvas.translate(toolPosition.x, toolPosition.y);
		canvas.rotate(rotationAngle);

		float textDescent = textPaint.descent();
		float textAscent = textPaint.ascent();

		float textHeight = textDescent - textAscent;
		float textBoxHeight = textHeight * multilineText.length + 2 * boxOffset;

		float maxTextWidth = 0;
		for (String str : multilineText) {
			float textWidth = textPaint.measureText(str);
			if (textWidth > maxTextWidth) {
				maxTextWidth = textWidth;
			}
		}
		float textBoxWidth = maxTextWidth + 2 * boxOffset;

		Bitmap textBitmap = Bitmap.createBitmap((int) textBoxWidth, (int) textBoxHeight,
				Bitmap.Config.ARGB_8888);
		Canvas textCanvas = new Canvas(textBitmap);

		for (int i = 0; i < multilineText.length; i++) {
			textCanvas.drawText(multilineText[i], boxOffset, boxOffset - textAscent + textHeight * i, textPaint);
		}

		Rect srcRect = new Rect(0, 0, (int) textBoxWidth, (int) textBoxHeight);
		Rect dstRect = new Rect((int) (-boxWidth / 2.0f), (int) (-boxHeight / 2.0f),
				(int) (boxWidth / 2.0f), (int) (boxHeight / 2.0f));
		canvas.drawBitmap(textBitmap, srcRect, dstRect, textPaint);

		canvas.restore();
	}

	@Override
	public void freeResources() {
	}
}

File: Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/TextToolCommand.kt
/*
 * Paintroid: An image manipulation application for Android.
 *  Copyright (C) 2010-2022 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.paintroid.command.implementation

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PointF
import org.catrobat.paintroid.command.Command
import org.catrobat.paintroid.command.serialization.SerializableTypeface
import org.catrobat.paintroid.common.ITALIC_FONT_BOX_ADJUSTMENT
import org.catrobat.paintroid.contract.LayerContracts
import org.catrobat.paintroid.tools.implementation.OUTLINED_FONT_WIDTH_ADJUSTMENT

const val TEXT_SIZE_MAGNIFICATION_FACTOR = 3f

class TextToolCommand(
    multilineText: Array<String>,
    textPaint: Paint,
    boxOffset: Float,
    boxWidth: Float,
    boxHeight: Float,
    toolPosition: PointF,
    rotationAngle: Float,
    typeFaceInfo: SerializableTypeface
) : Command {

    var multilineText = multilineText.clone(); private set
    var textPaint = textPaint; private set
    var boxOffset = boxOffset; private set
    var boxWidth = boxWidth; private set
    var boxHeight = boxHeight; private set
    var toolPosition = toolPosition; private set
    var rotationAngle = rotationAngle; private set
    var typeFaceInfo = typeFaceInfo; private set

    override fun run(canvas: Canvas, layerModel: LayerContracts.Model) {
        val textAscent = textPaint.ascent()
        val textDescent = textPaint.descent()
        val textHeight = (textDescent - textAscent) * multilineText.size
        val lineHeight = textHeight / multilineText.size
        var maxTextWidth = multilineText.maxOf { line ->
            textPaint.measureText(line)
        }

        if (typeFaceInfo.italic) {
            maxTextWidth *= ITALIC_FONT_BOX_ADJUSTMENT
        }

        with(canvas) {
            save()
            translate(toolPosition.x, toolPosition.y)
            rotate(rotationAngle)

            val widthScaling = (boxWidth - 2 * boxOffset) / maxTextWidth
            val heightScaling = (boxHeight - 2 * boxOffset) / textHeight
            canvas.scale(widthScaling, heightScaling)

            val scaledHeightOffset = boxOffset / heightScaling
            val scaledWidthOffset = boxOffset / widthScaling
            val scaledBoxWidth = boxWidth / widthScaling
            val scaledBoxHeight = boxHeight / heightScaling

            val fillPaint = Paint(textPaint)
            if (typeFaceInfo.outlined) fillPaint.color = Color.WHITE
            multilineText.forEachIndexed { index, textLine ->
                canvas.drawText(
                    textLine,
                    scaledWidthOffset - scaledBoxWidth / 2 / if (typeFaceInfo.italic) ITALIC_FONT_BOX_ADJUSTMENT else 1f,
                    -(scaledBoxHeight / 2) + scaledHeightOffset - textAscent + lineHeight * index,
                    fillPaint
                )
            }
            if (typeFaceInfo.outlined) {
                val outlinePaint = Paint(textPaint)
                val adjustedStrokeWidth = if (typeFaceInfo.outlineWidth == 0) 0f else java.lang.Float.max(textPaint.textSize / TEXT_SIZE_MAGNIFICATION_FACTOR * (typeFaceInfo.outlineWidth / OUTLINED_FONT_WIDTH_ADJUSTMENT), 1f)
                outlinePaint.style = Paint.Style.STROKE
                outlinePaint.strokeWidth = adjustedStrokeWidth
                multilineText.forEachIndexed { index, textLine ->
                    canvas.drawText(
                        textLine,
                        scaledWidthOffset - scaledBoxWidth / 2 / if (typeFaceInfo.italic) ITALIC_FONT_BOX_ADJUSTMENT else 1f,
                        -(scaledBoxHeight / 2) + scaledHeightOffset - textAscent + lineHeight * index,
                        outlinePaint
                    )
                }
            }
            restore()
        }
    }

    override fun freeResources() {
        // No resources to free
    }
}
